usaspending_api/download/helpers.py

```diff
@@ -99,62 +97,13 @@ def verify_requested_columns_available(sources, requested):
         raise InvalidParameterException('Unknown columns: {}'.format(bad_cols))
 
 
-# Multipart upload functions copied from Fabian Topfstedt's solution
-# http://www.topfstedt.de/python-parallel-s3-multipart-upload-with-retries.html
-def multipart_upload(bucketname, regionname, source_path, keyname, headers={}, guess_mimetype=True,
-                     parallel_processes=4):
-    """Parallel multipart upload."""
-    bucket = boto.s3.connect_to_region(regionname).get_bucket(bucketname)
-    if guess_mimetype:
-        mtype = mimetypes.guess_type(keyname)[0] or 'application/octet-stream'
-        headers.update({'Content-Type': mtype})
-
-    mp = bucket.initiate_multipart_upload(keyname, headers=headers)
-
+def multipart_upload(bucketname, regionname, source_path, keyname):
+    s3client = boto3.client('s3', region_name=regionname)
     source_size = os.stat(source_path).st_size
-    bytes_per_chunk = max(int(math.sqrt(5242880) * math.sqrt(source_size)),
-                          5242880)
-    chunk_amount = int(math.ceil(source_size / float(bytes_per_chunk)))
-
-    pool = multiprocessing.Pool(processes=parallel_processes)
-    for i in range(chunk_amount):
-        offset = i * bytes_per_chunk
-        remaining_bytes = source_size - offset
-        bytes = min([bytes_per_chunk, remaining_bytes])
-        part_num = i + 1
-        pool.apply_async(_upload_part, [bucketname, regionname, mp.id,
-                                        part_num, source_path, offset, bytes])
-    pool.close()
-    pool.join()
-
-    if len(mp.get_all_parts()) == chunk_amount:
-        mp.complete_upload()
-    else:
-        mp.cancel_upload()
-
-
-def _upload_part(bucketname, regionname, multipart_id, part_num, source_path, offset, bytes, amount_of_retries=10):
-    """Uploads a part with retries."""
-    bucket = boto.s3.connect_to_region(regionname).get_bucket(bucketname)
-
-    def _upload(retries_left=amount_of_retries):
-        try:
-            logging.info('Start uploading part #%d ...' % part_num)
-            for mp in bucket.get_all_multipart_uploads():
-                if mp.id == multipart_id:
-                    with FileChunkIO(source_path, 'r', offset=offset,
-                                     bytes=bytes) as fp:
-                        mp.upload_part_from_file(fp=fp, part_num=part_num)
-                    break
-        except Exception as exc:
-            if retries_left:
-                _upload(retries_left=retries_left - 1)
-            else:
-                logging.info('... Failed uploading part #%d' % part_num)
-                raise exc
-        else:
-            logging.info('... Uploaded part #%d' % part_num)
-    _upload()
+    bytes_per_chunk = max(int(math.sqrt(5242880) * math.sqrt(source_size)), 5242880)
```

Comment:
Great job using boto3 functions and replacing over 50 lines of convoluted (but functional) code.
Can you add comments here explaining the logic? What is the reason for using 5242880?

Reply:
It's a good question. boto3's `TransferConfig ` by default sets the multipart_threshold to 8 MB which I thought was pretty small given some of our files can be big, so I kept this bit thinking it optimized the part size. By the looks of it, it sets it from 5 MB to the square root of the filesize * 5 MB. I can arbitrarily set it to a constant though I'm not an expert on what to set it to.

Reply:
I think you meant to use `multipart_chunksize`, `multipart_threshold` is the filesize size before the file is chunked into multipart sections.

Reply:
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#boto3.s3.transfer.TransferConfig